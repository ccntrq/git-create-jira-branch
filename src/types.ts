// eslint-disable-next-line node/no-extraneous-import
import * as Schema from '@effect/schema/Schema';
import {Brand, Data, Match, Option, pipe} from 'effect';
import {dual} from 'effect/Function';

export type JiraApiUrl = string & Brand.Brand<'JiraApiUrl'>;
export const JiraApiUrl = Brand.nominal<JiraApiUrl>();
export type JiraPat = string & Brand.Brand<'JiraPat'>;
export const JiraPat = Brand.nominal<JiraPat>();
export type JiraApiToken = string & Brand.Brand<'JiraApiToken'>;
export const JiraApiToken = Brand.nominal<JiraApiToken>();
export type JiraUserEmail = string & Brand.Brand<'JiraUserEmail'>;
export const JiraUserEmail = Brand.nominal<JiraUserEmail>();
export type JiraKeyPrefix = string & Brand.Brand<'JiraKeyPrefix'>;
export const JiraKeyPrefix = Brand.nominal<JiraKeyPrefix>();

export type JiraCloudAuth = Data.Data<{
  _tag: 'JiraCloudAuth';
  jiraUserEmail: JiraUserEmail;
  jiraApiToken: JiraApiToken;
}>;

export const JiraCloudAuth = Data.tagged<JiraCloudAuth>('JiraCloudAuth');

export type JiraDataCenterAuth = Data.Data<{
  _tag: 'JiraDataCenterAuth';
  jiraPat: JiraPat;
}>;

export const JiraDataCenterAuth =
  Data.tagged<JiraDataCenterAuth>('JiraDataCenterAuth');

export type JiraAuth = JiraCloudAuth | JiraDataCenterAuth;

export type AppConfig = {
  jiraApiUrl: JiraApiUrl;
  defaultJiraKeyPrefix: Option.Option<JiraKeyPrefix>;
  jiraAuth: JiraAuth;
};

export const JiraIssuetypeSchema = Schema.struct({
  name: Schema.string,
});

export type JiraIssuetype = Schema.Schema.To<typeof JiraIssuetypeSchema>;

export const JiraIssueSchema = Schema.struct({
  key: Schema.string,
  fields: Schema.struct({
    summary: Schema.string,
    issuetype: JiraIssuetypeSchema,
  }),
});

export type GitCreateJiraBranchResult =
  | CreatedBranch
  | SwitchedBranch
  | ResetBranch;

export const matchGitCreateJiraBranchResult: {
  <A>(
    result: GitCreateJiraBranchResult,
    matcher: {
      onCreatedBranch: (branch: string) => A;
      onSwitchedBranch: (branch: string) => A;
      onResetBranch: (branch: string) => A;
    },
  ): A;
  <A>(matcher: {
    onCreatedBranch: (branch: string) => A;
    onSwitchedBranch: (branch: string) => A;
    onResetBranch: (branch: string) => A;
  }): (result: GitCreateJiraBranchResult) => A;
} = dual(
  2,
  <A>(
    result: GitCreateJiraBranchResult,
    matcher: {
      onCreatedBranch: (branch: string) => A;
      onSwitchedBranch: (branch: string) => A;
      onResetBranch: (branch: string) => A;
    },
  ): A =>
    pipe(
      Match.type<GitCreateJiraBranchResult>(),
      Match.tag('CreatedBranch', ({branch}) => matcher.onCreatedBranch(branch)),
      Match.tag('SwitchedBranch', ({branch}) =>
        matcher.onSwitchedBranch(branch),
      ),
      Match.tag('ResetBranch', ({branch}) => matcher.onResetBranch(branch)),
      Match.exhaustive,
    )(result) as A,
);

export interface CreatedBranch extends Data.Case {
  readonly _tag: 'CreatedBranch';
  readonly branch: string;
}

export const CreatedBranch = (branch: string): CreatedBranch =>
  Data.tagged<CreatedBranch>('CreatedBranch')({branch});

export interface SwitchedBranch extends Data.Case {
  readonly _tag: 'SwitchedBranch';
  readonly branch: string;
}

export const SwitchedBranch = (branch: string): SwitchedBranch =>
  Data.tagged<SwitchedBranch>('SwitchedBranch')({branch});

export interface ResetBranch extends Data.Case {
  readonly _tag: 'ResetBranch';
  readonly branch: string;
}

export const ResetBranch = (branch: string): ResetBranch =>
  Data.tagged<ResetBranch>('ResetBranch')({branch});

export type JiraIssue = Schema.Schema.To<typeof JiraIssueSchema>;

export type GitCreateJiraBranchError =
  | AppConfigError
  | GitExecError
  | JiraApiError;

export declare namespace GitCreateJiraBranchError {
  export interface Proto {
    readonly _tag: string;
    readonly message: string;
  }

  export type ProvidedFields = '_tag';
}

const makeError =
  <E extends GitCreateJiraBranchError.Proto>(tag: E['_tag']) =>
  (args: Omit<E, GitCreateJiraBranchError.ProvidedFields>): E =>
    Data.struct({
      _tag: tag,
      ...args,
    } as E);

export interface GitExecError extends GitCreateJiraBranchError.Proto {
  readonly _tag: 'GitExecError';
}

export const GitExecError = makeError<GitExecError>('GitExecError');

export interface AppConfigError extends GitCreateJiraBranchError.Proto {
  readonly _tag: 'AppConfigError';
}
export const AppConfigError = makeError<AppConfigError>('AppConfigError');

export interface JiraApiError extends GitCreateJiraBranchError.Proto {
  readonly _tag: 'JiraApiError';
}
export const JiraApiError = makeError<JiraApiError>('JiraApiError');
