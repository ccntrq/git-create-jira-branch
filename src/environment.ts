import {Config, ConfigError, Context, Effect, Layer} from 'effect';
import * as Option from 'effect/Option';

import {
  AppConfig,
  AppConfigError,
  JiraApiToken,
  JiraApiUrl,
  JiraCloudAuth,
  JiraDataCenterAuth,
  JiraKeyPrefix,
  JiraPat,
  JiraUserEmail,
} from './types';

export interface Environment {
  readonly getEnv: () => Effect.Effect<never, AppConfigError, AppConfig>;
}

export const Environment = Context.Tag<Environment>();

export const EnvironmentLive = Layer.succeed(
  Environment,
  Environment.of({
    getEnv: () =>
      Effect.all(
        [
          Config.string('JIRA_API_URL'),
          Config.option(Config.string('JIRA_PAT')),
          Config.option(Config.string('JIRA_USER_EMAIL')),
          Config.option(Config.string('JIRA_API_TOKEN')),
          Config.option(Config.string('JIRA_KEY_PREFIX')),
        ],
        {concurrency: 'unbounded', mode: 'validate'},
      ).pipe(
        Effect.catchAll((maybeErrors) => {
          return Effect.fail(
            AppConfigError({
              message: Option.reduceCompact<ConfigError.ConfigError, string[]>(
                maybeErrors,
                [],
                (acc, e) =>
                  acc.concat(
                    ConfigError.reduceWithContext(
                      e,
                      void 0,
                      mkCollectConfigErrorMessagesReducer(),
                    ),
                  ),
              ).join('\n'),
            }),
          );
        }),
        Effect.flatMap(
          ([
            jiraApiUrl,
            jiraPat,
            jiraUserEmail,
            jiraApiToken,
            defaultJiraKeyPrefix,
          ]) =>
            Effect.gen(function* (_) {
              if (
                Option.isSome(jiraPat) &&
                (Option.isSome(jiraUserEmail) || Option.isSome(jiraApiToken))
              ) {
                return yield* _(
                  Effect.fail(
                    AppConfigError({
                      message:
                        'Please provide either JIRA_PAT (for Jira Data Center) or JIRA_API_TOKEN and JIRA_USER_EMAIL (for Jira Cloud), but not both',
                    }),
                  ),
                );
              }

              if (Option.isSome(jiraPat)) {
                return {
                  jiraAuth: JiraDataCenterAuth({
                    jiraPat: JiraPat(jiraPat.value),
                  }),
                  jiraApiUrl: JiraApiUrl(jiraApiUrl),
                  defaultJiraKeyPrefix: Option.map(
                    defaultJiraKeyPrefix,
                    JiraKeyPrefix,
                  ),
                };
              }

              if (Option.isSome(jiraApiToken) && Option.isSome(jiraUserEmail)) {
                return {
                  jiraAuth: JiraCloudAuth({
                    jiraApiToken: JiraApiToken(jiraApiToken.value),
                    jiraUserEmail: JiraUserEmail(jiraUserEmail.value),
                  }),
                  jiraApiUrl: JiraApiUrl(jiraApiUrl),
                  defaultJiraKeyPrefix: Option.map(
                    defaultJiraKeyPrefix,
                    JiraKeyPrefix,
                  ),
                };
              }

              return yield* _(
                Effect.fail(
                  AppConfigError({
                    message:
                      'Please provide JIRA_PAT (for Jira Data Center) or JIRA_API_TOKEN and JIRA_USER_EMAIL (for Jira Cloud).',
                  }),
                ),
              );
            }),
        ),
      ),
  }),
);

const mkCollectConfigErrorMessagesReducer = (): ConfigError.ConfigErrorReducer<
  unknown,
  string[]
> => {
  const collectMessage = (
    _: unknown,
    _path: string[],
    message: string,
  ): string[] => [message];

  const mergeErrors = (
    _: unknown,
    left: string[],
    right: string[],
  ): string[] => [...left, ...right];

  return {
    andCase: mergeErrors,
    orCase: mergeErrors,
    invalidDataCase: collectMessage,
    missingDataCase: collectMessage,
    sourceUnavailableCase: collectMessage,
    unsupportedCase: collectMessage,
  };
};
