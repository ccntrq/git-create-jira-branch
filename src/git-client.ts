import { exec } from "child_process";
import { Context, Effect, Layer } from "effect";
import { GitExecError } from "./types";

type GitClientEffect<A> = Effect.Effect<never, GitExecError, A>;

export interface GitClient {
  readonly createGitBranch: (branchName: string) => GitClientEffect<void>;
  readonly createGitBranchFrom: (
    baseBranch: string
  ) => (branchName: string) => GitClientEffect<void>;
}

export const GitClient = Context.Tag<GitClient>();

const createGitBranchFrom =
  (baseBranch: string) =>
  (branchName: string): GitClientEffect<void> =>
    Effect.async((cb) => {
      exec(`git checkout -b ${branchName} ${baseBranch}`, (err) =>
        err !== null
          ? cb(
              Effect.fail(
                GitExecError({
                  message: `Checkout failed:\n'${err.message.trim()}'`,
                })
              )
            )
          : cb(Effect.succeed(undefined))
      );
    });

const gitClient = {
  createGitBranchFrom: createGitBranchFrom,
  createGitBranch: createGitBranchFrom(""),
};

export const GitClientLive = Layer.succeed(GitClient, GitClient.of(gitClient));
