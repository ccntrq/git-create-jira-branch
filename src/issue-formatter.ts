import type {JiraIssue} from './types';

import {Doc} from '@effect/printer';
import {bold, underlined} from '@effect/printer-ansi/Ansi';
import {render} from '@effect/printer-ansi/AnsiDoc';
import {pipe} from 'effect';

export const formatIssue = (issue: JiraIssue): string => {
  const heading = pipe(
    [
      mkBold(Doc.text(issue.key)),
      //Doc.text(issue.key),
      Doc.char('-'),
      Doc.text(issue.fields.summary),
    ],
    unwords,
    underline,
  );

  const details = pipe(
    [
      mkBold(Doc.text(issue.fields.issuetype.name)),
      Doc.char('|'),
      mkBold(Doc.text('Status:')),
      Doc.text(issue.fields.status.name),
      Doc.char('|'),
      mkBold(Doc.text('Creator:')),
      Doc.text(issue.fields.creator.displayName),
      ...(issue.fields.assignee
        ? [
            Doc.char('|'),
            mkBold(Doc.text('Assignee:')),
            Doc.text(issue.fields.assignee.displayName),
          ]
        : []),
    ],
    unwords,
    underline,
  );

  const description = issue.fields.description
    ? pipe(
        issue.fields.description.split(/\n|\r\n/),
        (d) =>
          // TODO:
          // use fillSep/reflow here to break lines at 80 characters when
          // performance issue is resolved
          // https://github.com/Effect-TS/effect/issues/2371
          // https://discord.com/channels/795981131316985866/1219736390401786008
          d.map(Doc.text),
        Doc.vcat,
      )
    : Doc.empty;

  const doc = pipe(
    [
      heading,
      details,
      ...(Doc.isEmpty(description) ? [] : [Doc.empty, description]),
    ],
    unlines,
  );

  return render(doc, {style: 'pretty'});
};

const unwords = <T>(docs: Iterable<Doc.Doc<T>>): Doc.Doc<T> =>
  Doc.concatWith<T>(Doc.catWithSpace)(docs);
const unlines = <T>(docs: Iterable<Doc.Doc<T>>): Doc.Doc<T> =>
  Doc.concatWith<T>(Doc.catWithLine)(docs);
const mkBold = Doc.annotate(bold);
const underline = Doc.annotate(underlined);
