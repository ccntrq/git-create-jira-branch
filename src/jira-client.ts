import { Layer, Context, Effect, pipe } from "effect";
import * as Http from "@effect/platform/HttpClient";
import * as ArrayFormatter from "@effect/schema/ArrayFormatter";
import * as ParseResult from "@effect/schema/ParseResult";

import { Environment } from "./environment";
import {
  AppConfigError,
  JiraApiError,
  JiraIssue,
  JiraIssueSchema,
} from "./types";

export interface JiraClient {
  readonly getJiraIssue: (
    issueKey: string
  ) => Effect.Effect<Environment, AppConfigError | JiraApiError, JiraIssue>;
}

export const JiraClient = Context.Tag<JiraClient>();

export const JiraClientLive = Layer.effect(
  JiraClient,
  Environment.pipe(
    Effect.map((env) =>
      JiraClient.of({
        getJiraIssue: (issueId: string) => {
          const endPoint = `/rest/api/latest/issue/${issueId}`;

          return env.getEnv().pipe(
            Effect.flatMap(({ jiraPat, jiraApiUrl }) =>
              Http.request
                .get(jiraApiUrl + endPoint, {
                  headers: {
                    Authorization: `Bearer ${jiraPat}`,
                    Accept: "application/json",
                  },
                })
                .pipe(Http.client.fetchOk())
            ),
            Effect.flatMap(Http.response.schemaBodyJson(JiraIssueSchema)),
            handleJiraClientErrors
          );
        },
      })
    )
  )
);

const handleJiraClientErrors = (
  eff: Effect.Effect<
    Environment,
    AppConfigError | ParseResult.ParseError | Http.error.HttpClientError,
    JiraIssue
  >
): Effect.Effect<Environment, AppConfigError | JiraApiError, JiraIssue> =>
  eff.pipe(
    Effect.catchTag("ParseError", (e) =>
      Effect.fail(
        JiraApiError({
          message: [
            "Failed to parse ticket response from Jira:",
            ...ArrayFormatter.formatErrors(e.errors).map(
              (issue) => `'${issue.path.join(" ")}': '${issue.message}'`
            ),
          ].join("\n"),
        })
      )
    ),
    Effect.catchTag("RequestError", (e) =>
      Effect.fail(
        JiraApiError({
          message: `Failed to make ticket request to Jira. Reason: ${e.reason}`,
        })
      )
    ),
    Effect.catchTag("ResponseError", (e) =>
      Effect.fail(
        JiraApiError({
          message: `Jira Ticket request returned failure. Reason: ${e.reason}${
            typeof e.error === "string" ? ` (${e.error})` : ""
          } StatusCode: ${e.response.status}`,
        })
      )
    )
  );
