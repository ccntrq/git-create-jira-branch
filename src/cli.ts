import {Option, Data, Effect, pipe, Console} from 'effect';
import {ValidationError} from '@effect/cli';
import {CliApp} from '@effect/cli/CliApp';
import {isNone} from 'effect/Option';
import * as Args from '@effect/cli/Args';
import * as Command from '@effect/cli/Command';
import * as Options from '@effect/cli/Options';
import * as Span from '@effect/cli/HelpDoc/Span';
import * as HelpDoc from '@effect/cli/HelpDoc';
import * as CommandExecutor from '@effect/platform/CommandExecutor';

import {
  gitCreateJiraBranch,
  ticketUrl,
  ticketUrlForCurrentBranch,
} from './core';
import * as packageJson from '../package.json';
import {
  GitCreateJiraBranchError,
  matchGitCreateJiraBranchResult,
} from './types';
import {compose} from 'effect/Function';
import {AppConfigService} from './app-config';
import {GitClient} from './git-client';
import {JiraClient} from './jira-client';
import {openUrl} from './url-opener';

interface GitCreateJiraBranch {
  readonly jiraKey: Option.Option<string>;
  readonly baseBranch: Option.Option<string>;
  readonly reset: boolean;
  readonly open: boolean;
}

const GitCreateJiraBranch = Data.case<GitCreateJiraBranch>();

const mainCommand = pipe(
  Command.make(
    'git-create-jira-branch',
    {
      options: Options.all({
        baseBranch: Options.withDescription(
          Options.optional(Options.withAlias(Options.text('baseBranch'), 'b')),
          'Base branch to create the new branch from',
        ),
        reset: Options.withDescription(
          Options.withAlias(Options.boolean('reset'), 'r'),
          'Reset the branch if it already exists',
        ),
        open: Options.withDescription(
          Options.withAlias(Options.boolean('open'), 'o'),
          'Open the ticket for the current branch or the given ticket in the browser',
        ),
      }),
      args: Args.withDescription(
        Args.atMost(Args.text({name: 'jira-key'}), 1),
        'The Jira ticket key to create a branch for (e.g. FOOX-1234)',
      ),
    },
    (
      args,
    ): Effect.Effect<
      void,
      GitCreateJiraBranchError,
      | AppConfigService
      | GitClient
      | JiraClient
      | CommandExecutor.CommandExecutor
    > => {
      const command = GitCreateJiraBranch({
        baseBranch: args.options.baseBranch,
        reset: args.options.reset,
        open: args.options.open,
        jiraKey: Option.fromNullable(args.args[0]),
      });

      if (command.open) {
        return pipe(
          command.jiraKey,
          Option.match({
            onSome: (jiraKey) => ticketUrl(jiraKey),
            onNone: () => ticketUrlForCurrentBranch(),
          }),
          Effect.tap((url) =>
            Console.log(
              `Opening ticket url '${url}' in your default browser...`,
            ),
          ),
          Effect.flatMap(openUrl),
        );
      }

      if (isNone(command.jiraKey)) {
        return printDocs(HelpDoc.p(Span.error('No Jira Key provided')));
      }

      return Effect.flatMap(
        gitCreateJiraBranch(
          command.jiraKey.value,
          command.baseBranch,
          command.reset,
        ),
        compose(
          matchGitCreateJiraBranchResult({
            onCreatedBranch: (branch) =>
              `Successfully created branch: '${branch}'`,
            onSwitchedBranch: (branch) =>
              `Switched to already existing branch: '${branch}'`,
            onResetBranch: (branch) => `Reset branch: '${branch}'`,
          }),
          Console.log,
        ),
      );
    },
  ),
  Command.withDescription(
    HelpDoc.p(
      `Fetches the given Jira ticket and creates an aproriately named branch for it.
The branch type (bug or feat) is determined by the ticket type. The branch name
is based on the ticket summary.`,
    ),
  ),
);

const cli = {
  name: packageJson.name,
  version: packageJson.version,
  summary: Span.text(packageJson.description),
};

export const cliEffect = (
  args: string[],
): Effect.Effect<
  void,
  never,
  | CliApp.Environment
  | GitClient
  | AppConfigService
  | JiraClient
  | CommandExecutor.CommandExecutor
> =>
  Command.run(
    mainCommand,
    cli,
  )(args).pipe(
    Effect.catchIf(ValidationError.isValidationError, (e) =>
      printDocs(HelpDoc.p(Span.error(HelpDoc.getSpan(e.error)))),
    ),
    Effect.catchAll((e) => printDocs(HelpDoc.p(Span.error(e.message)))),
  );

const printDocs = (doc: HelpDoc.HelpDoc): Effect.Effect<void> =>
  Console.log(HelpDoc.toAnsiText(doc));
