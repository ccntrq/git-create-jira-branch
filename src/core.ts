import {Chunk, Effect, Option} from 'effect';
import {isNone} from 'effect/Option';

import {GitClient} from './git-client';
import {JiraClient} from './jira-client';
import {Environment} from './environment';
import {
  CreatedBranch,
  GitCreateJiraBranchError,
  GitCreateJiraBranchResult,
  JiraIssue,
  JiraIssuetype,
  JiraKeyPrefix,
  SwitchedBranch,
} from './types';

export const gitCreateJiraBranch = (
  jiraKey: string,
  baseBranch: Option.Option<string>,
): Effect.Effect<
  Environment | GitClient | JiraClient,
  GitCreateJiraBranchError,
  GitCreateJiraBranchResult
> =>
  Effect.gen(function* ($) {
    const [envProvider, gitClient, jiraClient] = yield* $(
      Effect.all([Environment, GitClient, JiraClient]),
    );

    const {defaultJiraKeyPrefix} = yield* $(envProvider.getEnv());
    const fullKey = buildJiraKey(defaultJiraKeyPrefix, jiraKey);

    const issue = yield* $(jiraClient.getJiraIssue(fullKey));
    const branchName = jiraIssueToBranchName(issue);

    const branchExists = yield* $(
      Effect.map(gitClient.listBranches(), Chunk.contains(branchName)),
    );

    if (branchExists) {
      yield* $(gitClient.switchBranch(branchName));
      return SwitchedBranch(branchName);
    }

    yield* $(
      Option.match(baseBranch, {
        onNone: () => gitClient.createGitBranch,
        onSome: gitClient.createGitBranchFrom,
      })(branchName),
    );

    return CreatedBranch(branchName);
  });

const slugify = (str: string): string =>
  str
    .toLowerCase()
    .replace(/ä/g, 'ae')
    .replace(/ö/g, 'oe')
    .replace(/ü/g, 'ue')
    .replace(/ß/g, 'ss')
    .replace(/[^a-zA-Z0-9]/g, '-')
    .replace(/-+/g, '-')
    .replace(/^-|-$/g, '');

const jiraIssueToBranchName = (issue: JiraIssue): string => {
  const branchtype = jiraIssuetypeBranchtype(issue.fields.issuetype);
  return `${branchtype}/${issue.key}-${slugify(issue.fields.summary)}`;
};

const jiraIssuetypeBranchtype = (issuetype: JiraIssuetype): string => {
  if (issuetype.name.match(/bug/i)) {
    return 'fix';
  }

  return 'feat';
};

const buildJiraKey = (
  defaultJiraKeyPrefix: Option.Option<JiraKeyPrefix>,
  jiraKey: string,
): string =>
  jiraKey.match(/^([a-z]+)-(\d+)$/i) || isNone(defaultJiraKeyPrefix)
    ? jiraKey
    : `${defaultJiraKeyPrefix.value}-${jiraKey}`;
