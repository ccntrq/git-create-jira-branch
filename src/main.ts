#!/usr/bin/env node

import * as NodeCommandExecutor from '@effect/platform-node/NodeCommandExecutor';
import * as NodeFileSystem from '@effect/platform-node/NodeFileSystem';
import * as Http from '@effect/platform/HttpClient';
import {Cause, Console, Effect, Layer, pipe} from 'effect';

import {NodeContext} from '@effect/platform-node';
import {AppConfigService} from './app-config';
import {cliEffect} from './cli';
import {GitClientLive} from './git-client';
import {JiraClientLive} from './jira-client';

const commandExecutorLayer = NodeCommandExecutor.layer.pipe(
  Layer.provide(NodeFileSystem.layer),
);

const gitClientLayer = GitClientLive.pipe(Layer.provide(commandExecutorLayer));

const jiraClientLayer = JiraClientLive.pipe(
  Layer.provide(Http.client.layer),
  Layer.provide(AppConfigService.Live),
);

const mainLive = Layer.mergeAll(
  commandExecutorLayer,
  gitClientLayer,
  AppConfigService.Live,
  jiraClientLayer,
  NodeContext.layer,
);

const mainEffect = pipe(
  Effect.sync(() => process.argv),
  Effect.flatMap(cliEffect),
  Effect.provide(mainLive),
  Effect.catchAllDefect((defect) => {
    if (Cause.isRuntimeException(defect)) {
      return Console.log(`RuntimeException defect caught: ${defect.message}`);
    }
    return Console.log(`Unknown defect caught: ${JSON.stringify(defect)}`);
  }),
);

Effect.runPromise(mainEffect);
