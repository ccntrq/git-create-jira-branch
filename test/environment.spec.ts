import {Environment, EnvironmentLive} from '../src/environment';
import {ConfigProvider, Effect, Either, Layer} from 'effect';
import * as Option from 'effect/Option';
import {describe, expect} from 'vitest';
import {itEffect} from './util';
import {
  JiraApiToken,
  JiraCloudAuth,
  JiraDataCenterAuth,
  JiraPat,
  JiraUserEmail,
} from '../src/types';

const testProg = Effect.all([Environment]).pipe(
  Effect.flatMap(([env]) => env.getEnv()),
);

const mkTestLayer = (
  configProvider: ConfigProvider.ConfigProvider,
): Layer.Layer<never, never, Environment> =>
  EnvironmentLive.pipe(Layer.provide(Layer.setConfigProvider(configProvider)));

describe('Environment', () => {
  itEffect('should provide jira data center appconfig', () =>
    Effect.gen(function* ($) {
      const configProvider = ConfigProvider.fromMap(
        new Map([
          ['JIRA_PAT', 'dummy-jira-pat'],
          ['JIRA_API_URL', 'https://dummy-jira-instance.com'],
          ['JIRA_KEY_PREFIX', 'DUMMYAPP'],
        ]),
      );

      const env = yield* $(
        Effect.provide(testProg, mkTestLayer(configProvider)),
      );

      expect(env).toEqual({
        defaultJiraKeyPrefix: Option.some('DUMMYAPP'),
        jiraApiUrl: 'https://dummy-jira-instance.com',
        jiraAuth: JiraDataCenterAuth({
          jiraPat: JiraPat('dummy-jira-pat'),
        }),
      });
    }),
  );

  itEffect('should provide jira cloud appconfig', () =>
    Effect.gen(function* ($) {
      const configProvider = ConfigProvider.fromMap(
        new Map([
          ['JIRA_API_TOKEN', 'dummy-jira-api-token'],
          ['JIRA_USER_EMAIL', 'dummyuser@example.com'],
          ['JIRA_API_URL', 'https://dummy-jira-instance.com'],
          ['JIRA_KEY_PREFIX', 'DUMMYAPP'],
        ]),
      );

      const env = yield* $(
        Effect.provide(testProg, mkTestLayer(configProvider)),
      );

      expect(env).toEqual({
        defaultJiraKeyPrefix: Option.some('DUMMYAPP'),
        jiraApiUrl: 'https://dummy-jira-instance.com',
        jiraAuth: JiraCloudAuth({
          jiraApiToken: JiraApiToken('dummy-jira-api-token'),
          jiraUserEmail: JiraUserEmail('dummyuser@example.com'),
        }),
      });
    }),
  );

  itEffect('should provide appconfig with missing optional values', () =>
    Effect.gen(function* ($) {
      const configProvider = ConfigProvider.fromMap(
        new Map([
          ['JIRA_PAT', 'dummy-jira-pat'],
          ['JIRA_API_URL', 'https://dummy-jira-instance.com'],
        ]),
      );

      const env = yield* $(
        Effect.provide(testProg, mkTestLayer(configProvider)),
      );

      expect(env).toEqual({
        defaultJiraKeyPrefix: Option.none(),
        jiraApiUrl: 'https://dummy-jira-instance.com',
        jiraAuth: JiraDataCenterAuth({
          jiraPat: JiraPat('dummy-jira-pat'),
        }),
      });
    }),
  );

  itEffect('should report missing configuration values', () =>
    Effect.gen(function* ($) {
      const configProvider = ConfigProvider.fromMap(
        new Map([['JIRA_API_URL', 'https://dummy-jira-instance.com']]),
      );

      const result = yield* $(
        Effect.either(Effect.provide(testProg, mkTestLayer(configProvider))),
      );
      Either.match(result, {
        onLeft: (e) => expect(e).toMatchSnapshot(),
        onRight: (_) => expect.unreachable('Should have returned an error'),
      });
    }),
  );
});
