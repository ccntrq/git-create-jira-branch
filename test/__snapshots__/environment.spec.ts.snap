// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Environment > should report missing configuration values 1`] = `
{
  "_tag": "AppConfigError",
  "message": "Please provide JIRA_PAT (for Jira Data Center) or JIRA_API_TOKEN and JIRA_USER_EMAIL (for Jira Cloud).",
}
`;
